src/riak_cs_bw_limitter.erl

```diff
@@ -0,0 +1,197 @@
+%%
+%% Copyright (c) 2007-2015 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% ---------------------------------------------------------------------
+
+%% @doc Simple node-level rate limit implementation, per user
+%% This is experimental feature.
+%%
+%% riak_cs_bw_limitter help
+%%
+%%    This module enables node-level rate limiting.
+%%
+%%  set_limits(DurationSec, BandwidthMax, AccessCountMax)
+%%
+%%    Set bandwidth and access limit. DurationSec is refresh rate.
+%%    set_limits(-1, -1, -1) will remove all rate limitation.
+%%
+%%  state()
+%%
+%%    Shows cached list of users being limited.
+%%    {user_state, User Key, Current bytes, Access count}.
+%%
+%%  reset()
+%%
+%%    Resets current state rate limiting. Limits won't be cleared.
+
+-module(riak_cs_bw_limitter).
+
+-behaviour(riak_cs_quota).
+
+-include("riak_cs.hrl").
+-include_lib("webmachine/include/webmachine_logger.hrl").
+
+-export([state/0, reset/0, set_limits/3]).
+
+-export([init/0, allow/4, update/3, handle_info/2,
+         update_state/2, terminate/1]).
+-record(state, {
+          table :: atom(),
+          %% Default values. -1 means no quota.
+          duration_sec = -1 :: integer(), %% Refresh rate in secs
+          bandwidth_max = -1 :: integer(), %% allowed Bytes per second
+          access_count_max = -1 :: integer() %% allowed access per sercond
+         }).
+
+-record(user_state, {
+          user :: binary(),
+          bandwidth = 0 :: non_neg_integer(), %% Bytes per refresh rate
+          access_count = 0 :: non_neg_integer() %% Access count per refresh rate
+         }).
+
+-spec set_limits(integer(), integer(), integer()) -> ok.
+set_limits(DurationSec, BandwidthMax, AccessCountMax) ->
+    riak_cs_quota:update(?MODULE,
+                         {DurationSec, BandwidthMax, AccessCountMax}).
+
+-spec state() -> [#user_state{}].
+state() ->
+    {ok, #state{table=Table}} = riak_cs_quota:get_local_state(?MODULE),
+    ets:tab2list(Table).
+
+-spec reset() -> true.
+reset() ->
+    {ok, #state{table=Table}} = riak_cs_quota:get_local_state(?MODULE),
+    ets:delete_all_objects(Table).
+
+-spec init() -> {ok, #state{}}.
+init() ->
+    %% TableName = list_to_atom(lists:flatten([atom_to_list(?MODULE), $_, "user"])),
+    TableName = ?MODULE,
+    Table = ets:new(TableName, [{write_concurrency, true}, {keypos, 2}, public, set]),
```

Comment:
I prefer `#user_state.user` instead of `2` for readability. It explains this table stores `#user_state` records.